File: module-7-evals/solutions/graders/code-grader.ts
/**
 * Module 7: Deterministic code graders
 * String match, file exists, test passes, static analysis.
 */
import * as fs from "fs/promises";
import * as path from "path";
import { execSync } from "child_process";
import type { Grader, GradeResult, Transcript } from "../types.js";

/** Check if a file contains an expected string */
export function stringMatchGrader(filePath: string, expected: string): Grader {
  return {
    name: `string-match(${filePath})`,
    grade: async (workspaceDir): Promise<GradeResult> => {
      try {
        const content = await fs.readFile(path.join(workspaceDir, filePath), "utf-8");
        const found = content.includes(expected);
        return {
          score: found ? 1 : 0,
          passed: found,
          explanation: found
            ? `Found expected string in ${filePath}`
            : `Expected string not found in ${filePath}`,
        };
      } catch {
        return { score: 0, passed: false, explanation: `File ${filePath} not found` };
      }
    },
  };
}

/** Check if a file exists */
export function fileExistsGrader(filePath: string): Grader {
  return {
    name: `file-exists(${filePath})`,
    grade: async (workspaceDir): Promise<GradeResult> => {
      try {
        await fs.access(path.join(workspaceDir, filePath));
        return { score: 1, passed: true, explanation: `File ${filePath} exists` };
      } catch {
        return { score: 0, passed: false, explanation: `File ${filePath} not found` };
      }
    },
  };
}

/** Run a shell command and check exit code */
export function shellTestGrader(command: string, name?: string): Grader {
  return {
    name: name || `shell-test(${command.slice(0, 40)})`,
    grade: async (workspaceDir): Promise<GradeResult> => {
      try {
        const output = execSync(command, {
          cwd: workspaceDir,
          encoding: "utf-8",
          timeout: 30000,
        });
        return { score: 1, passed: true, explanation: `Command passed: ${output.trim().slice(0, 200)}` };
      } catch (e) {
        return {
          score: 0,
          passed: false,
          explanation: `Command failed: ${e instanceof Error ? e.message.slice(0, 200) : String(e)}`,
        };
      }
    },
  };
}

/** Composite grader: all sub-graders must pass (partial credit) */
export function compositeGrader(graders: Grader[]): Grader {
  return {
    name: `composite(${graders.map((g) => g.name).join(", ")})`,
    grade: async (workspaceDir, transcript): Promise<GradeResult> => {
      const results = await Promise.all(graders.map((g) => g.grade(workspaceDir, transcript)));
      const avgScore = results.reduce((sum, r) => sum + r.score, 0) / results.length;
      const allPassed = results.every((r) => r.passed);
      const explanation = results.map((r) => `${r.passed ? "✓" : "✗"} ${r.explanation}`).join("\n");
      return { score: avgScore, passed: allPassed, explanation };
    },
  };
}

File: module-7-evals/solutions/python/graders/code_grader.py
"""Module 7: Deterministic code graders."""
import os
import subprocess
import sys

sys.path.insert(0, os.path.join(os.path.dirname(__file__), ".."))
from eval_types import GradeResult, Grader, Transcript


def string_match_grader(file_path: str, expected: str) -> Grader:
    def grade(workspace_dir: str, transcript: Transcript = None) -> GradeResult:
        try:
            with open(os.path.join(workspace_dir, file_path)) as f:
                content = f.read()
            found = expected in content
            return GradeResult(
                score=1.0 if found else 0.0,
                passed=found,
                explanation=f"Found expected string in {file_path}" if found else f"Expected string not found in {file_path}",
            )
        except FileNotFoundError:
            return GradeResult(score=0.0, passed=False, explanation=f"File {file_path} not found")

    return Grader(name=f"string-match({file_path})", grade=grade)


def file_exists_grader(file_path: str) -> Grader:
    def grade(workspace_dir: str, transcript: Transcript = None) -> GradeResult:
        exists = os.path.exists(os.path.join(workspace_dir, file_path))
        return GradeResult(
            score=1.0 if exists else 0.0,
            passed=exists,
            explanation=f"File {file_path} exists" if exists else f"File {file_path} not found",
        )

    return Grader(name=f"file-exists({file_path})", grade=grade)


def shell_test_grader(command: str, name: str = None) -> Grader:
    def grade(workspace_dir: str, transcript: Transcript = None) -> GradeResult:
        try:
            result = subprocess.run(
                command, shell=True, capture_output=True, text=True,
                cwd=workspace_dir, timeout=30,
            )
            if result.returncode == 0:
                return GradeResult(score=1.0, passed=True, explanation=f"Command passed: {result.stdout.strip()[:200]}")
            return GradeResult(score=0.0, passed=False, explanation=f"Command failed: {result.stderr.strip()[:200]}")
        except Exception as e:
            return GradeResult(score=0.0, passed=False, explanation=f"Command failed: {e}")

    return Grader(name=name or f"shell-test({command[:40]})", grade=grade)


def composite_grader(graders: list[Grader]) -> Grader:
    def grade(workspace_dir: str, transcript: Transcript = None) -> GradeResult:
        results = [g.grade(workspace_dir, transcript) for g in graders]
        avg_score = sum(r.score for r in results) / len(results)
        all_passed = all(r.passed for r in results)
        explanation = "\n".join(f"{'✓' if r.passed else '✗'} {r.explanation}" for r in results)
        return GradeResult(score=avg_score, passed=all_passed, explanation=explanation)

    return Grader(name=f"composite({', '.join(g.name for g in graders)})", grade=grade)
